[Bundle 4/n] Prune g_chainman usage in validation-adjacent modules

dongcarl opened the pull request:
Overall PR: #20158 (tree-wide: De-globalize ChainstateManager)

Based on:
- [x] #21055 | [Bundle 3/n] Prune g_chainman usage in mempool-related validation functions

Note to reviewers:
1. This bundle may _apparently_ introduce usage of `g_chainman` or `::Chain(state|)Active()` globals, but these are resolved later on in the overall PR. [Commits of overall PR](https://github.com/bitcoin/bitcoin/pull/20158/commits)
2. There may be seemingly obvious local references to `ChainstateManager` or other validation objects which are not being used in callers of the current function in question, this is done intentionally to **_keep each commit centered around one function/method_** to ease review and to make the overall change systematic. We don't assume anything about our callers. Rest assured that once we are considering that particular caller in later commits, we will use the obvious local references. [Commits of overall PR](https://github.com/bitcoin/bitcoin/pull/20158/commits)
3. When changing a function/method that has many callers (e.g. `LookupBlockIndex` with 55 callers), it is sometimes easier (and less error-prone) to use a scripted-diff. When doing so, there will be 3 commits in sequence so that every commit compiles like so:
	1. Add `new_function`, make `old_function` a wrapper of `new_function`, divert all calls to `old_function` to `new_function` **in the local module only**
	2. Scripted-diff to divert all calls to `old_function` to `new_function` **in the rest of the codebase**
	3. Remove `old_function`

dongcarl committed:
node/coinstats: Pass in BlockManager to GetUTXOStats

dongcarl committed:
net_processing: Move some static functions to PeerManager

- BlockRequestAllowed
- AlreadyHaveBlock
- ProcessGetBlockData
- PrepareBlockFilterRequest
- ProcessGetCFilters
- ProcessGetCFHeaders
- ProcessGetCFCheckPt

Moved out of anonymous namespace:
- ProcessBlockAvailability
- UpdateBlockAvailability
- CanDirectFetch

dongcarl committed:
miner: Pass in blockman to ::RegenerateCommitments

REQUIRES ATTENTION

dongcarl committed:
node: Use existing NodeContext

dongcarl committed:
scripted-diff: net_processing: Use existing chainman

-BEGIN VERIFY SCRIPT-
sed -i -E \
    -e 's/g_chainman/m_chainman/g' \
    -e 's@([^:])(Chain(state|)Active)@\1::\2@g' \
    -e 's@::Chain(state|)Active\(\)@m_chainman.ActiveChain\1()@g' \
    -- src/net_processing.cpp
-END VERIFY SCRIPT-

dongcarl committed:
net_processing: Add review-only assertion to PeerManager

dongcarl committed:
miner: Pass in chainstate to BlockAssembler::CreateNewBlock

dongcarl committed:
scripted-diff: Invoke CreateNewBlock with chainstate

-BEGIN VERIFY SCRIPT-
find_regex='(\.|->)CreateNewBlock\(' \
    && git grep -l -E "$find_regex" -- src \
        | grep -v '^src/miner\.\(cpp\|h\)$' \
        | xargs sed -i -E 's@'"$find_regex"'@\0::ChainstateActive(), @g'
-END VERIFY SCRIPT-

dongcarl committed:
node/ifaces: NodeImpl: Use existing NodeContext member

dongcarl committed:
node/ifaces: ChainImpl: Use existing NodeContext member

dongcarl committed:
miner: Remove old CreateNewBlock w/o chainstate param

ryanofsky reviewed (approved):
Code review ACK last 11 commits only a0741fa4d5acb567122d0d43c725f8f64fd75eaf. Changes since last review in https://github.com/bitcoin/bitcoin/pull/20158#pullrequestreview-581753681: rebase, and a few more functions moved in move static functions commit

- [X] c8382b98a493635ffe57ad5daa5e03fdecdb3b77 miner: Pass in chainstate to BlockAssembler::CreateNewBlock (17/27)
- [X] c625552140ce8b9845f657a04963573a98daaba5 scripted-diff: Invoke CreateNewBlock with chainstate (18/27)
- [X] b186b2f2129dcfac52fb2dea9230c6cf29a952c2 miner: Remove old CreateNewBlock w/o chainstate param (19/27)
- [X] 588a0c75d7b139dc572866dd92de7177adfb5136 miner: Pass in blockman to ::RegenerateCommitments (20/27)
- [X] c46ff650d3b55f2e5a1de5a28a4946aedc357368 node/coinstats: Pass in BlockManager to GetUTXOStats (21/27)
- [X] 62ab369eacb4d39fe74df12697b8a410636116c9 node: Use existing NodeContext (22/27)
- [X] c975f64b7e742105ab3cfb10596fdbaaba8bec7a node/ifaces: NodeImpl: Use existing NodeContext member (23/27)
- [X] 6723fe829672cd6b5487e690852aff180d8a4106 node/ifaces: ChainImpl: Use existing NodeContext member (24/27)
- [X] 15276edbbee67941d1483504eace147dcfbf25ec net_processing: Move some static functions to PeerManager (25/27)
- [X] 0ac24f921ebe33027466ec9aaf75201fef5622db scripted-diff: net_processing: Use existing chainman (26/27)
- [X] a0741fa4d5acb567122d0d43c725f8f64fd75eaf net_processing: Add review-only assertion to PeerManager (27/27)

jnewbery commented:
Concept ACK. Will review after rebase.

dongcarl commented:
Pushed a0741fa4d5acb567122d0d43c725f8f64fd75eaf -> 38c325b1df093b2ff380df93c3a518286c531f32
- Rebased over master

dongcarl committed:
validation: Remove extraneous LoadGenesisBlock function prototype

Leftover from last bundle.

dongcarl commented:
Pushed 38c325b1df093b2ff380df93c3a518286c531f32 -> a67983cd6d8e61565da4e03f3ba401d0148fe195
- Addressed https://github.com/bitcoin/bitcoin/pull/21270#commitcomment-47985104 in prepended commit
- Fixed namespace brace syntax mistake in "net_processing: Move some static functions to PeerManager"

ryanofsky reviewed (approved):
Code review ACK a67983cd6d8e61565da4e03f3ba401d0148fe195. Only change since last review new first commit fixing header declaration, and rebase

glozow reviewed (commented):
code review ACK a67983cd6d8e61565da4e03f3ba401d0148fe195

Reviewed commits individually. It helped to `git show 021a04a --color-moved=dimmed-zebra` on the _net_processing: Move some static functions to PeerManager_ commit.

Question: I forget if this was already mentioned somewhere, but why isn't `LookupBlockIndex` a `const` member function? Otherwise could be passing `const BlockManager&`

laanwj commented on src/miner.cpp:
```diff
@@ -39,13 +39,13 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
     block.vtx.at(0) = MakeTransactionRef(tx);
 
-    GenerateCoinbaseCommitment(block, WITH_LOCK(cs_main, return g_chainman.m_blockman.LookupBlockIndex(block.hashPrevBlock)), Params().GetConsensus());
+    GenerateCoinbaseCommitment(block, WITH_LOCK(::cs_main, assert(std::addressof(g_chainman.m_blockman) == std::addressof(blockman)); return blockman.LookupBlockIndex(block.hashPrevBlock)), Params().GetConsensus());
```
This is a really strange C++ construction to me, to have assertions and return statements inside a function argument. It is likely out of scope of this PR to refactor it though. I'm not sure splitting it up over multiple lines makes it much more insightful.

MarcoFalke replied:
The assert will go away eventually, if I understood correctly, so the line is only temporarily longer than it previously was

laanwj replied:
Right :smile:

ghostofjnewbery replied:
I had a heart attack, but I've returned from beyond the grave to relay this warning from the underworld. Do not make 150 column function calls containing asserts and return statements, lest one day you should have to read and understand that code.

Consider doing this instead. It's a lot less spooky:

```suggestion
    CBlockIndex* index;
    {
        LOCK(::cs_main);
        assert(std::addressof(g_chainman.m_blockman) == std::addressof(blockman));
        index = blockman.LookupBlockIndex(block.hashPrevBlock);
    }
    GenerateCoinbaseCommitment(block, index, Params().GetConsensus());
```

laanwj commented:
Code review ACK a67983cd6d8e61565da4e03f3ba401d0148fe195

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1920,27 +1939,27 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, const Peer& peer,
         // are still present, however, as belt-and-suspenders.
 
-        if (received_new_header && pindexLast->nChainWork > ::ChainActive().Tip()->nChainWork) {
+        if (received_new_header && pindexLast->nChainWork > m_chainman.ActiveChain().Tip()->nChainWork) {
```
This sequence of `m_chainman.ActiveChain()` function calls is repeatedly taking a `cs_main` lock and releasing it again (while holding `cs_main` the whole time). That's fine, and correct, but perhaps it'd be better to just take a local `const CChainstate&` variable once to avoid the multiple function calls.

jnewbery commented:
> Question: I forget if this was already mentioned somewhere, but why isn't LookupBlockIndex a const member function? Otherwise could be passing const BlockManager&

I can't see any reason why `LookupBlockIndex` shouldn't be a const function. Changing all the places that `BlockManager&` is passed to pass `const BlockManager&` seems like it'd be a big improvement and indicate to the reader that the function won't mutate the `BlockManager`.

jnewbery reviewed (commented):
Looks good to me. A few comments inline for your consideration for follow-up PRs.

It seems a shame that net_processing and other components need to make so many calls into `ActiveChainstate()` and `ActiveChain()`, simply to call `IsInitialBlockDownload()`, `ActivateBestChain()`, or another global function, or pass the received pointer back to a validation function. It suggests to me that the interface we have between net_processing and validation isn't quite right here. net_processing really shouldn't have to know which chainstate is active for most of its purposes.

MarcoFalke commented on src/miner.cpp:
```diff
@@ -99,7 +99,7 @@ void BlockAssembler::resetBlock()
 Optional<int64_t> BlockAssembler::m_last_block_weight{nullopt};
 
-std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& scriptPubKeyIn)
+std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(CChainState& chainstate, const CScript& scriptPubKeyIn)
```
This should really be a member variable, like the mempool. I'd doubt it would make sense to pass a different chainstate reference than the one the mempool is based on.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -472,6 +472,24 @@ class PeerManagerImpl final : public PeerManager
+
+    void ProcessBlockAvailability(NodeId nodeid) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    void UpdateBlockAvailability(NodeId nodeid, const uint256 &hash) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    bool CanDirectFetch(const Consensus::Params &consensusParams) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
Isn't params a member? Seems odd to pass twice

MarcoFalke replied:
#21425

MarcoFalke commented on src/node/coin.cpp:
```diff
@@ -12,7 +12,8 @@ void FindCoins(const NodeContext& node, std::map<COutPoint, Coin>& coins)
     LOCK2(cs_main, node.mempool->cs);
-    CCoinsViewCache& chain_view = ::ChainstateActive().CoinsTip();
+    assert(std::addressof(::ChainstateActive()) == std::addressof(node.chainman->ActiveChainstate()));
+    CCoinsViewCache& chain_view = node.chainman->ActiveChainstate().CoinsTip();
```
Any reason to prefer UB over a clean assert, like it is done for node.mempool above?

MarcoFalke commented on src/node/interfaces.cpp:
```diff
@@ -182,18 +182,21 @@ class NodeImpl : public Node
         LOCK(::cs_main);
-        return ::ChainActive().Height();
+        assert(std::addressof(::ChainActive()) == std::addressof(m_context->chainman->ActiveChain()));
+        return m_context->chainman->ActiveChain().Height();
```
same

MarcoFalke commented on src/node/transaction.cpp:
```diff
@@ -39,9 +39,10 @@ TransactionError BroadcastTransaction(NodeContext& node, const CTransactionRef t
     // If the transaction is already confirmed in the chain, don't do anything
     // and return early.
-    CCoinsViewCache &view = ::ChainstateActive().CoinsTip();
+    CCoinsViewCache &view = node.chainman->ActiveChainstate().CoinsTip();
```
same

dongcarl commented:
Thanks everyone for the diligent reviews! I will prepend commits in Bundle 5 when I get the chance and link back to the convo here, we can continue the convos over on the Bundle 5 thread!

fanquake commented on src/node/coinstats.h:
```diff
@@ -8,6 +8,7 @@
 
 #include <amount.h>
 #include <uint256.h>
+#include <validation.h>
```
Done in #21525.

fanquake commented on src/net_processing.cpp:
```diff
@@ -684,41 +702,6 @@ bool PeerManagerImpl::MarkBlockAsInFlight(NodeId nodeid, const uint256& hash, co
     return true;
 }
 
-/** Check whether the last unknown block a peer advertised is not yet known. */
```
Done in #21525
